8Knot/pages/codebase/visualizations/reviewer_file_heatmap.py

```diff
@@ -0,0 +1,484 @@
+    logging.warning(f"DIRECTORY DROPDOWN - FINISHED")
+
+    return directories, "Top Level Directory"
+
+
+# callback for reviewer file heatmap graph
+@callback(
+    Output(f"{PAGE}-{VIZ_ID}", "figure"),
+    [
+        Input(f"repo-{PAGE}-{VIZ_ID}", "value"),
+        Input(f"directory-{PAGE}-{VIZ_ID}", "value"),
+        Input("bot-switch", "value"),
+    ],
+    background=True,
+)
+def reviewer_file_heatmap_graph(repo_id, directory, bot_switch):
+    start = time.perf_counter()
+    logging.warning(f"{VIZ_ID}- START")
+
+    # get dataframes of data from cache
+    df_file, df_actions, df_file_cntbs = multi_query_helper([repo_id])
+
+    # test if there is data
+    if df_file.empty or df_actions.empty or df_file_cntbs.empty:
+        logging.warning(f"{VIZ_ID} - NO DATA AVAILABLE")
+        return nodata_graph
+
+    # function for all data pre processing
+    df = process_data(df_file, df_actions, df_file_cntbs, directory, bot_switch)
+
+    # if there are no cntrbs in a directory plot no data graph
+    if df.empty:
+        return nodata_graph
+
+    fig = create_figure(df)
+
+    logging.warning(f"{VIZ_ID} - END - {time.perf_counter() - start}")
+    return fig
+
+
+def multi_query_helper(repos):
+    """
+    For reviewer_file_heatmap_graph-
+    hack to put all of the cache-retrieval
+    in the same place temporarily
+    """
+
+    # wait for data to asynchronously download and become available.
+    while not_cached := cf.get_uncached(func_name=rfq.__name__, repolist=repos):
+        logging.warning(f"CONTRIBUTOR FILE HEATMAP - WAITING ON DATA TO BECOME AVAILABLE")
+        time.sleep(0.5)
+
+    # wait for data to asynchronously download and become available.
+    while not_cached := cf.get_uncached(func_name=cnq.__name__, repolist=repos):
+        logging.warning(f"CONTRIBUTOR FILE HEATMAP - WAITING ON DATA TO BECOME AVAILABLE")
+        time.sleep(0.5)
+
+    # wait for data to asynchronously download and become available.
+    while not_cached := cf.get_uncached(func_name=cpfq.__name__, repolist=repos):
+        logging.warning(f"CONTRIBUTOR FILE HEATMAP - WAITING ON DATA TO BECOME AVAILABLE")
+        time.sleep(0.5)
+
+    # GET ALL DATA FROM POSTGRES CACHE
+    df_file = cf.retrieve_from_cache(
+        tablename=rfq.__name__,
+        repolist=repos,
+    )
+    df_actions = cf.retrieve_from_cache(
+        tablename=cnq.__name__,
+        repolist=repos,
+    )
+    df_file_cntrbs = cf.retrieve_from_cache(
+        tablename=cpfq.__name__,
+        repolist=repos,
+    )
+
+    # necessary preprocessing steps that were lifted out of the querying step
+    df_actions = preproc_u.contributors_df_action_naming(df_actions)
+    df_file_cntrbs = preproc_u.cntrb_per_file(df_file_cntrbs)
+
+    return df_file, df_actions, df_file_cntrbs
+
+
+def process_data(
+    df_file: pd.DataFrame,
+    df_actions: pd.DataFrame,
+    df_file_cntbs: pd.DataFrame,
+    directory,
+    bot_switch,
+):
+
```

Comment:
Verbal overview of processing steps would be useful here- contextualizes following functions. 

e.g.

1) Relate files in the repository to the contributors who have reviewed them in past PRs. Clean up formatting.
2) For a given level in the directory tree, aggregate the list of contributors for sub-directories and for individual files at the level.
3) For each contributor, identify their most recent contribution.
4) Relate (file/subdirectory:reviewers) mapping done in step (2) to contributor history done in step (3) to derive a monthly count of reviewers last-seen.

Reply:
This isn't an ideal verbal breakdown, but I think it illustrates what I'd love to have